modules/Nncase.Modules.CPU/Passes/CPUFunctionPartitionPass.cs

```diff
@@ -23,85 +26,198 @@ protected override Task<IRModule> RunCoreAsync(IRModule module, RunPassContext c
-
-#if Debug
-                ctx.GraphSummary.DumpDot(DumpScope.Current.Directory + $"function_{i}_summary.dot");
-#endif
-
-                var dfsVisitor = new QuikGraph.Algorithms.TopologicalSort.SourceFirstTopologicalSortAlgorithm<Vertex, Edge>(ctx.GraphSummary);
-                dfsVisitor.Compute();
-                var exprMemo = new Dictionary<Expr, Expr>(ReferenceEqualityComparer.Instance);
-                for (var vi = 0; vi < dfsVisitor.SortedVertices.Length; vi++)
+                    return true;
+                }
+
+                return f.CheckedType is DistributedType;
+            }
+
+            bool isSupport = false;
+            switch (arg.Edge.Source.Expr, arg.Edge.Target.Expr)
+            {
+                case (Call callee, Call caller):
+                    switch (callee.CheckedType, caller.CheckedType)
+                    {
+                        case (DistributedType, TensorType) when caller.Target is IR.CPU.Boxing:
+                        case (DistributedType, DistributedType):
+                            isSupport = true;
+                            break;
+                    }
+
+                    break;
+                case (IR.Tuple tpArg, Call caller):
+                    if (tpArg.Fields.AsValueEnumerable().All(f => f.CheckedType is DistributedType) && caller.CheckedType is DistributedType)
+                    {
+                        isSupport = true;
+                    }
+
+                    break;
+                case (Call field, IR.Tuple tp):
+                    isSupport = tp.Fields.AsValueEnumerable().All(f => f is Call c && CheckField(c)) ? true : false;
+                    break;
+                default:
+                    break;
+            }
+
+            return isSupport;
+        };
+
+        condenseAlgo.IsGraphCompatible += (algo, edge) =>
+        {
+            return algo.CondensedGraph.IsDirectedAcyclicGraph();
+        };
+
+        condenseAlgo.Compute();
+
+        if (Diagnostics.DumpScope.Current.IsEnabled(Diagnostics.DumpFlags.Rewrite))
+        {
+            condenseAlgo.CondensedGraph.Dump($"{funcName}Condensed", init => { });
+            condenseAlgo.ClusteredGraph.Dump($"{funcName}Cluster", algo =>
+            {
+                algo.FormatVertex += (s, arg) =>
                 {
-                    var vertex = dfsVisitor.SortedVertices[vi];
-                    var subgraph = ctx.SubgraphMap[ctx.SummaryVertexSubgraphMap[vertex]];
-                    if (vertex.CompatType == Compat.INCOMPATIBLE)
+                    arg.VertexFormat.Label = $"{arg.Vertex.Expr.GetType().Name}";
+                };
+            });
+        }
+
+        // 3. reconstruction
+        var constructor = new DistributedReConstructor(funcName, ModuleKind, condenseAlgo);
+        var post = constructor.Construct();
+        return post;
+    }
+}
+
+internal sealed class DistributedReConstructor : ExprReConstructor<ExprVertex, ExprEdge>
```

Comment:
`ReConstructor` should be spelled "Reconstructor".

Reply:
got it. it will be fixed in next pr.